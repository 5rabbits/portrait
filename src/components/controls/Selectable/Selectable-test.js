import { React, toJSON, mount } from 'test/helper'
import cx from 'classnames'
import Selectable from './Selectable'

describe('Selectable', () => {
  it('should render with minimum props', () => {
    const component = mount(
      <Selectable
        options={[]}
        renderer={() => <span />}
      />,
    )

    expect(toJSON(component)).toMatchSnapshot()
  })

  it('should allow to render options', () => {
    const component = mount(
      <Selectable
        options={[
          { label: 'Option 1', value: 1 },
          { label: 'Option 10', value: 10 },
          { label: 'Option 5', value: 5 },
        ]}
        renderer={
          ({
            options,
          }) =>
            <div>
              {options.map(option =>
                <div key={option.value}>{option.label}</div>,
              )}
            </div>
        }
      />,
    )

    expect(toJSON(component)).toMatchSnapshot()
  })

  it('should sort options', () => {
    const renderer = jest.fn().mockImplementation(() => <span />)

    mount(
      <Selectable
        options={[
          { label: 'Option 1', value: 1 },
          { label: 'Option 5', value: 5 },
          { label: 'Option 10', value: 10 },
        ]}
        renderer={renderer}
      />,
    )

    expect(renderer.mock.calls[0][0].options).toEqual([
      { label: 'Option 1', value: 1 },
      { label: 'Option 10', value: 10 },
      { label: 'Option 5', value: 5 },
    ])
  })

  describe('focusable elements', () => {
    it('should allow to define focusable elements', () => {
      const component = mount(
        <Selectable
          options={[
            { label: 'Option 1', value: 1 },
            { label: 'Option 10', value: 10 },
            { label: 'Option 5', value: 5 },
          ]}
          renderer={
            ({
              focusableRef,
              focusedElement,
              options,
            }) =>
              <div>
                {options.map((option, index) =>
                  <div
                    className={cx({
                      focused: focusedElement === index,
                    })}
                    key={option.value}
                    ref={focusableRef(index)}
                    >
                    {option.label}
                  </div>,
                )}
              </div>
          }
        />,
      )
      const instance = component.instance()

      instance.setFocusedElement(2)

      expect(component.find('.focused').text()).toBe('Option 5')
    })

    it('should allow to give real focus to registered elements', () => {
      const component = mount(
        <Selectable
          options={[
            { label: 'Option 1', value: 1 },
            { label: 'Option 10', value: 10 },
            { label: 'Option 5', value: 5 },
          ]}
          renderer={
            ({
              focusableRef,
              options,
            }) =>
              <div>
                <input ref={focusableRef('input')} />

                {options.map(option =>
                  <div key={option.value}>{option.label}</div>,
                )}
              </div>
          }
        />,
      )
      const instance = component.instance()

      instance.setFocusedElement('input', { virtual: false })

      expect(component.find('input').getDOMNode()).toBe(document.activeElement)
    })

    describe('if the element don\'t respond to focus', () => {
      it('should not fail', () => {
        class Example extends React.PureComponent {
          render() {
            return <span />
          }
        }

        const component = mount(
          <Selectable
            options={[
              { label: 'Option 1', value: 1 },
              { label: 'Option 10', value: 10 },
              { label: 'Option 5', value: 5 },
            ]}
            renderer={
              ({
                focusableRef,
                options,
              }) =>
                <div>
                  <Example ref={focusableRef('nonFocusable')} />

                  {options.map(option =>
                    <div key={option.value}>{option.label}</div>,
                  )}
                </div>
            }
          />,
        )
        const instance = component.instance()

        expect(() => {
          instance.setFocusedElement('nonFocusable', { virtual: false })
        }).not.toThrow()
      })
    })

    it('should cycle the focus when is a number', () => {
      const component = mount(
        <Selectable
          options={[
            { label: 'Option 1', value: 1 },
            { label: 'Option 5', value: 5 },
            { label: 'Option 10', value: 10 },
          ]}
          renderer={() => <span />}
        />,
      )
      const instance = component.instance()

      instance.setFocusedElement(-1)

      expect(instance.getFocusedElement()).toBe(2)

      instance.setFocusedElement(4)

      expect(instance.getFocusedElement()).toBe(1)
    })
  })

  describe('searching', () => {
    it('should allow to filter options', () => {
      const renderer = jest.fn().mockImplementation(() => <span />)

      mount(
        <Selectable
          options={[
            { label: 'Option 1', value: 1 },
            { label: 'Option 5', value: 5 },
            { label: 'Option 10', value: 10 },
          ]}
          renderer={renderer}
          search="1"
        />,
      )

      expect(renderer.mock.calls[0][0].options).toEqual([
        { label: 'Option 1', value: 1 },
        { label: 'Option 10', value: 10 },
      ])
    })

    it('should expose a method to get search matches', () => {
      const renderer = jest.fn().mockImplementation(() => <span />)

      mount(
        <Selectable
          options={[
            { label: 'Option 1', value: 1 },
            { label: 'Option 5', value: 5 },
            { label: 'Option 10', value: 10 },
          ]}
          renderer={renderer}
          search="10"
        />,
      )

      expect(renderer.mock.calls[0][0].getSearchMatches('Option 10')).toEqual([
        { fromIndex: 0, match: false, text: 'Option ', toIndex: 7 },
        { fromIndex: 7, match: true, text: '10', toIndex: 9 },
      ])
    })

    describe('if there is no search', () => {
      it('should returna default text match', () => {
        const renderer = jest.fn().mockImplementation(() => <span />)

        mount(
          <Selectable
            options={[
              { label: 'Option 1', value: 1 },
              { label: 'Option 5', value: 5 },
              { label: 'Option 10', value: 10 },
            ]}
            renderer={renderer}
          />,
        )

        expect(renderer.mock.calls[0][0].getSearchMatches('Option 10')).toEqual([
          { fromIndex: 0, match: false, text: 'Option 10', toIndex: 8 },
        ])
      })
    })
  })

  describe('when clicking outside', () => {
    it('should invoke props.onClickOutside', () => {
      const onClickOutside = jest.fn()
      const target = document.createElement('div')
      const container = document.createElement('div')

      document.body.appendChild(target)
      document.body.appendChild(container)

      mount(
        <Selectable
          onClickOutside={onClickOutside}
          options={[]}
          renderer={() => <span />}
        />,
        { attachTo: container },
      )

      target.dispatchEvent(new CustomEvent('click', { bubbles: true }))

      expect(onClickOutside).toHaveBeenCalled()
    })

    it('should remove focus if props.blurOnClickOutside', () => {
      const target = document.createElement('div')
      const container = document.createElement('div')

      document.body.appendChild(target)
      document.body.appendChild(container)

      const component = mount(
        <Selectable
          blurOnClickOutside
          options={[]}
          renderer={() => <span />}
        />,
        { attachTo: container },
      )
      const instance = component.instance()

      instance.setFocused(true)

      expect(instance.isFocused()).toBe(true)

      target.dispatchEvent(new CustomEvent('click', { bubbles: true }))

      expect(instance.isFocused()).toBe(false)
    })

    it('should stop listening when unmounting', () => {
      const onClickOutside = jest.fn()
      const target = document.createElement('div')
      const container = document.createElement('div')

      document.body.appendChild(target)
      document.body.appendChild(container)

      const component = mount(
        <Selectable
          onClickOutside={onClickOutside}
          options={[]}
          renderer={() => <span />}
        />,
        { attachTo: container },
      )

      component.unmount()
      target.dispatchEvent(new CustomEvent('click', { bubbles: true }))

      expect(onClickOutside).not.toHaveBeenCalled()
    })
  })

  describe('overflow area', () => {
    it('should set a relative position on the overflow container if not specified', () => {
      const component = mount(
        <Selectable
          options={[]}
          renderer={({ overflowRef }) => <div ref={overflowRef()} />}
        />,
      )

      expect(component.find('div').getDOMNode().style.position).toBe('relative')
    })

    it('should set a relative position on the overflow container if is static', () => {
      const component = mount(
        <Selectable
          options={[]}
          renderer={({ overflowRef }) => <div ref={overflowRef()} style={{ position: 'static' }} />}
        />,
      )

      expect(component.find('div').getDOMNode().style.position).toBe('relative')
    })

    it('should not fail when unmounting', () => {
      const component = mount(
        <Selectable
          options={[]}
          renderer={({ overflowRef }) => <div ref={overflowRef()} />}
        />,
      )

      expect(() => {
        component.unmount()
      }).not.toThrow()
    })
  })

  describe('when receiving new props', () => {
    describe('if props.options changes', () => {
      it('should sort the new options', () => {
        const component = mount(
          <Selectable
            options={[]}
            renderer={() => <span />}
          />,
        )
        const instance = component.instance()

        component.setProps({
          options: [
            { label: 'Option 1', value: 1 },
            { label: 'Option 5', value: 5 },
            { label: 'Option 10', value: 10 },
          ],
        })

        expect(instance.getVisibleOptions()).toEqual([
          { label: 'Option 1', value: 1 },
          { label: 'Option 10', value: 10 },
          { label: 'Option 5', value: 5 },
        ])
      })
    })

    describe('if props.value changes', () => {
      it('should set the newly selected option', () => {
        const component = mount(
          <Selectable
            options={[
              { label: 'Option 1', value: 1 },
              { label: 'Option 5', value: 5 },
            ]}
            renderer={() => <span />}
            value={1}
          />,
        )
        const instance = component.instance()

        component.setProps({ value: 5 })

        expect(instance.getSelectedOption()).toEqual({ label: 'Option 5', value: 5 })

        component.setProps({ value: null })

        expect(instance.getSelectedOption()).toBe(null)
      })
    })
  })

  it('should expose search helpers', () => {
    const component = mount(
      <Selectable
        options={[]}
        renderer={() => <span />}
      />,
    )
    const instance = component.instance()

    instance.setSearch('test')

    expect(instance.getSearch()).toBe('test')
  })

  it('should expose value helpers', () => {
    const component = mount(
      <Selectable
        options={[]}
        renderer={() => <span />}
      />,
    )
    const instance = component.instance()

    instance.setValue(2)

    expect(instance.getValue()).toBe(2)
  })

  it('should expose focus helpers', () => {
    const component = mount(
      <Selectable
        options={[]}
        renderer={() => <span />}
      />,
    )
    const instance = component.instance()

    instance.focus()

    expect(instance.isFocused()).toBe(true)

    instance.blur()

    expect(instance.isFocused()).toBe(false)
  })

  describe('if trying to move the scroll of a node outside the overflow container', () => {
    it('should indicate that is visible', () => {
      const component = mount(
        <Selectable
          options={[]}
          renderer={({ overflowRef }) => <div ref={overflowRef()} />}
        />,
      )
      const instance = component.instance().getWrappedInstance()
      const node = document.createElement('div')

      expect(instance.isNodeInViewport(node).inViewport).toBe(true)
    })
  })

  describe('if trying to move the scroll on a node inside the overflow container', () => {
    it('should calculate if it is in the viewport', () => {
      const component = mount(
        <Selectable
          options={[]}
          renderer={({ focusableRef, overflowRef }) =>
            <div ref={overflowRef()}>
              <div ref={focusableRef(0)} />
            </div>
          }
        />,
      )
      const instance = component.instance().getWrappedInstance()
      const info = instance.isNodeInViewport(instance.getFocusableNode(0))

      expect(Object.keys(info)).toEqual([
        'inViewport', 'visibleFromBottom', 'visibleFromTop',
      ])
    })
  })

  describe('scrolling to focused', () => {
    describe('if the element is hidden on the bottom the overflow', () => {
      it('should scroll to node to the bottom', () => {
        const component = mount(
          <Selectable
            options={[
              { label: 'Option 1', value: 1 },
            ]}
            renderer={({ focusableRef, overflowRef }) =>
              <div ref={overflowRef()}>
                <div ref={focusableRef(0)} />
              </div>
            }
          />,
        )
        const instance = component.instance().getWrappedInstance()

        instance.focusableRefs[0] = {}
        instance.isNodeInViewport = () => ({ visibleFromBottom: false })
        instance.setFocusedElement(0)

        expect(() => {
          instance.scrollToFocusedElement()
        }).not.toThrow()
      })
    })

    describe('if the element is hidden on the top the overflow', () => {
      it('should scroll to node to the bottom', () => {
        const component = mount(
          <Selectable
            options={[
              { label: 'Option 1', value: 1 },
            ]}
            renderer={({ focusableRef, overflowRef }) =>
              <div ref={overflowRef()}>
                <div ref={focusableRef(0)} />
              </div>
            }
          />,
        )
        const instance = component.instance().getWrappedInstance()

        instance.focusableRefs[0] = {}
        instance.isNodeInViewport = () => ({ visibleFromBottom: true, visibleFromTop: false })
        instance.setFocusedElement(0)

        expect(() => {
          instance.scrollToFocusedElement()
        }).not.toThrow()
      })
    })

    describe('if is entirely visible', () => {
      it('should not try to scrollNodeToViewport', () => {
        const component = mount(
          <Selectable
            options={[
              { label: 'Option 1', value: 1 },
            ]}
            renderer={({ focusableRef, overflowRef }) =>
              <div ref={overflowRef()}>
                <div ref={focusableRef(0)} />
              </div>
            }
          />,
        )
        const instance = component.instance().getWrappedInstance()

        instance.scrollNodeToViewport = jest.fn()
        instance.isNodeInViewport = () => ({ visibleFromBottom: true, visibleFromTop: true })
        instance.setFocusedElement(0)
        instance.scrollToFocusedElement()

        expect(instance.scrollNodeToViewport).not.toHaveBeenCalled()
      })
    })

    describe('if there is no overflow defined', () => {
      it('should not try to scrollNodeToViewport', () => {
        const component = mount(
          <Selectable
            options={[
              { label: 'Option 1', value: 1 },
            ]}
            renderer={({ focusableRef }) =>
              <div>
                <div ref={focusableRef(0)} />
              </div>
            }
          />,
        )
        const instance = component.instance().getWrappedInstance()

        instance.scrollNodeToViewport = jest.fn()
        instance.setFocusedElement(0)
        instance.scrollToFocusedElement()

        expect(instance.scrollNodeToViewport).not.toHaveBeenCalled()
      })
    })
  })
})
