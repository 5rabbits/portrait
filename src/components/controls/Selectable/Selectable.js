import { PureComponent } from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import deburr from 'lodash/deburr'
import sortBy from 'lodash/sortBy'
import isNumber from 'lodash/isNumber'
import controllable from 'helpers/controllable'

@controllable({
  focused: 'onFocusedChange',
  focusedElement: 'onFocusedElementChange',
  search: 'onSearchChange',
  value: 'onChange',
}, ['scrollToFocusedElement'])
export default class Selectable extends PureComponent {
  static propTypes = {
    // Handled by uncontrollable
    /* eslint-disable react/no-unused-prop-types */
    defaultFocused: PropTypes.bool,
    defaultFocusedElement: PropTypes.any,
    defaultSearch: PropTypes.any,
    defaultValue: PropTypes.string,
    /* eslint-enable react/no-unused-prop-types */

    blurOnClickOutside: PropTypes.bool,
    focused: PropTypes.bool.isRequired,
    focusedElement: PropTypes.any,
    onChange: PropTypes.func.isRequired,
    onClickOutside: PropTypes.func,
    onFocusedChange: PropTypes.func.isRequired,
    onFocusedElementChange: PropTypes.func.isRequired,
    onSearchChange: PropTypes.func.isRequired,
    options: PropTypes.array.isRequired,
    renderer: PropTypes.func.isRequired,
    search: PropTypes.string,
    value: PropTypes.any,
  }

  static defaultProps = {
    blurOnClickOutside: true,
    defaultFocused: false,
    defaultSearch: '',
  }

  constructor(props) {
    super(props)

    const sortedOptions = this.sortOptions(this.props.options)
    let selectedOption

    if (this.props.value != null) {
      selectedOption = sortedOptions.find(option => option.value === this.props.value)
    }

    this.state = {
      focusedElement: this.props.focusedElement,
      options: this.filterOptions({
        options: sortedOptions,
        search: this.props.search,
      }),
      selectedOption,
      sortedOptions,
    }

    this.focusableRefs = {}
  }

  componentDidMount() {
    this.container = ReactDOM.findDOMNode(this) // eslint-disable-line react/no-find-dom-node
    document.addEventListener('click', this.handleOutsideClick)
  }

  componentWillReceiveProps(nextProps) {
    let sortedOptions = this.state.sortedOptions

    if (this.props.options !== nextProps.options) {
      sortedOptions = this.sortOptions(nextProps.options)
      this.setState({ sortedOptions })
    }

    if (
      this.props.options !== nextProps.options ||
      this.props.search !== nextProps.search
    ) {
      this.setState({
        options: this.filterOptions({
          options: sortedOptions,
          search: nextProps.search,
        }),
      })
    }

    if (this.props.value !== nextProps.value) {
      this.setState({
        selectedOption: nextProps.value == null
          ? null
          : nextProps.options.find(option => option.value === nextProps.value),
      })
    }

    if (
      this.props.search !== nextProps.search ||
      this.props.focused !== nextProps.focused
    ) {
      nextProps.onFocusedElementChange(null)
    }

    if (this.props.focusedElement !== nextProps.focusedElement) {
      let elementToFocus = nextProps.focusedElement

      if (isNumber(nextProps.focusedElement)) {
        const optionsCount = this.state.options.length

        elementToFocus = nextProps.focusedElement < 0
          ? (optionsCount + nextProps.focusedElement) % optionsCount
          : nextProps.focusedElement % optionsCount
      }

      this.setState({ focusedElement: elementToFocus })
    }
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleOutsideClick)
  }

  getFocusableElement = id => (
    this.focusableRefs[id]
  )

  getRendererProps = () => {
    const { focused, search, value } = this.props
    const { focusedElement, options, selectedOption } = this.state

    return {
      focusableRef: this.focusableRef,
      focusedElement,
      getSearchMatches: this.searchMatches,
      focused,
      options,
      overflowRef: this.overflowRef,
      search,
      selectedOption,
      setFocused: this.props.onFocusedChange,
      setFocusedElement: this.props.onFocusedElementChange,
      setSearch: this.props.onSearchChange,
      setValue: this.props.onChange,
      scrollToFocusedElement: this.scrollToFocusedElement,
      value,
    }
  }

  isNodeInViewport(node) {
    if (!this.overflow.contains(node)) {
      return {
        inViewport: true,
        visibleFromBottom: true,
        visibleFromTop: true,
      }
    }

    const overflow = this.overflow
    const nodeTop = node.offsetTop
    const visibleFromTop = nodeTop >= overflow.scrollTop
    const visibleFromBottom = nodeTop + node.offsetHeight <=
      overflow.scrollTop + overflow.offsetHeight

    return {
      inViewport: visibleFromBottom && visibleFromTop,
      visibleFromBottom,
      visibleFromTop,
    }
  }

  scrollToFocusedElement = () => {
    requestAnimationFrame(() => {
      const { focusedElement } = this.state
      const node = this.focusableRefs[focusedElement]

      if (this.overflow && focusedElement != null) {
        const viewportInfo = this.isNodeInViewport(node)

        if (!viewportInfo.visibleFromBottom) {
          this.scrollNodeToViewport(node, 'bottom')
        }
        else if (!viewportInfo.visibleFromTop) {
          this.scrollNodeToViewport(node, 'top')
        }
      }
    })
  }

  scrollNodeToViewport = (node, fromDirection) => {
    const scrollTop = fromDirection === 'bottom'
      ? node.offsetTop - (this.overflow.offsetHeight - node.offsetHeight)
      : node.offsetTop

    this.overflow.scrollTop = scrollTop
  }

  filterOptions = ({ options, search }) => {
    if (!search.trim()) {
      return options
    }

    const words = deburr(search.trim()).split(/\s+/)
    const pattern = words.map(word => `(?=.*${word})`).join('')
    const regexp = new RegExp(pattern, 'ig')

    return options.filter(option => regexp.test(deburr(option.label)))
  }

  searchMatches = text => {
    const { search } = this.props

    if (!search) {
      return [{
        text,
        match: false,
        fromIndex: 0,
        toIndex: text.length - 1,
      }]
    }

    const pattern = `(${deburr(search.trim()).split(/\s+/).join('|')})`
    const regexp = new RegExp(pattern, 'ig')
    const terms = []
    let currentIndex = 0

    deburr(text).split(regexp).forEach(term => {
      if (term === '') {
        return
      }

      const fromIndex = currentIndex
      const toIndex = currentIndex + term.length
      const termText = text.substring(fromIndex, toIndex)
      const match = !!term.match(regexp)

      currentIndex = toIndex

      terms.push({
        text: termText,
        match,
        fromIndex,
        toIndex,
      })
    })

    return terms
  }

  sortOptions = options => (
    sortBy(options, option => deburr(option.label.toLowerCase()))
  )

  handleOutsideClick = event => {
    const isOutside = (
      this.container &&
      this.container !== event.target &&
      !this.container.contains(event.target) &&
      event.target.parentNode
    )

    if (isOutside && this.props.blurOnClickOutside && this.props.focused) {
      this.props.onFocusedChange(false)
    }

    if (isOutside && this.props.onClickOutside) {
      this.props.onClickOutside(this.getRendererProps())
    }
  }

  focusableRef = id => node => {
    this.focusableRefs[id] = node
  }

  overflowRef = () => overflow => {
    this.overflow = overflow

    if (overflow && (!overflow.style.position || overflow.style.position === 'static')) {
      overflow.style.position = 'relative' // eslint-disable-line no-param-reassign
    }
  }

  render() {
    return this.props.renderer(this.getRendererProps())
  }
}
