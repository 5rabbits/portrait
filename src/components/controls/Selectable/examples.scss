@import '~core/variables';

.component-example--Selectable {
  .selectable-dropdown {
    background-color: #fff;
    position: relative;
    width: 200px;
  }

  .selectable-dropdown__control {
    width: 100%;
    outline: none;
    border: #cdd4db solid 1px;
    outline: none;
    text-align: left;
    height: 32px;
    background-color: #fff;
    padding: 0 12px;
    position: relative;
    z-index: 1;
    margin: 0;
  }

  .selectable-dropdown__menu {
    position: absolute;
    right: 0;
    left: 0;
    z-index: 99999;
    background-color: #fff;
    background-clip: padding-box;
    box-shadow: rgba(0, 0, 0, 0.35) 0 8px 24px;
    border: rgba(0, 0, 0, 0.1) solid 1px;
    margin-top: -1px;
  }

  .selectable-dropdown__option {
    height: 32px;
    width: 100%;
    background: none;
    border: none;
    outline: none;
    text-align: left;
    margin: 0;
    padding: 0 12px;
    display: flex;
    align-items: center;
  }

  .selectable-dropdown__option--focused {
    background-color: #eee;
  }

  .selectable-dropdown__option--selected {
    background-color: $primary-color;
    color: #fff;
  }

  .selectable-dropdown__searchWrapper {
    position: relative;
  }

  .selectable-dropdown__searchIcon {
    border: none;
    outline: none;
    background: none;
    margin: 0;
    position: absolute;
    top: 50%;
    right: 8px;
    transform: translateY(-50%);
  }

  .selectable-dropdown__searchInput {
    width: 100%;
    height: 32px;
    padding: 0 32px 0 12px;
    background-color: #FFF;
    border: none;
    border-bottom: $default-input-border-color solid 1px;

    &::placeholderÂ {
      color: $placeholder-color;
    }
  }

  .selectable-dropdown__overflow {
    max-height: 112px;
    overflow-y: auto;
  }

  .selectable-dropdown__bottomOption {
    height: 40px;
    width: 100%;
    background: none;
    border: none;
    outline: none;
    text-align: left;
    margin: 0;
    padding: 0 12px;
    border-top: $default-input-border-color solid 1px;
    color: $placeholder-color;
  }

  .selectable-dropdown--without-value {
    .selectable-dropdown__control {
      color: $placeholder-color;
    }
  }

  .current-value {
    margin-top: 16px;
  }
}
