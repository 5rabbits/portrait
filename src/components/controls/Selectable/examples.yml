-
  description: |
    A simple selectable options list. It uses these minimum arguments:

    - `getOptionProps(option, otherProps = {})`: Get props for an option.
    - `options`: Sorted options array.
    - `value`: The current value.

    Its important to note that both `getContainerProps` and `getOptionProps` allows to
    declare additional props for the component without worrying about overriding core
    props declared by Selectable.

  code: >
    <Selectable
      onChange={console.log}
      options={[
        { label: 'Daniel Espinoza', value: 2 },
        { label: 'Víctor Sánchez', value: 5 },
        { label: 'Tamara Luque', value: 4 },
        { label: 'Abraham Barrera', value: 1 },
        { label: 'Rodrigo Díaz', value: 3 },
      ]}
      renderer={
        ({
          getOptionProps,
          options,
          value,
        }) =>
          <div className="selectable-dropdown">
            {options.map(option =>
              <button
                {...getOptionProps(option, {
                  className: classNames('selectable-dropdown__option', {
                    'selectable-dropdown__option--selected': value === option.value,
                  }),
                  type: 'button',
                })}
                >
                {option.label}
              </button>
            )}
          </div>
      }
    />

-
  description: |
    Adds a search input using:

    - `getSearchMatches(text)`: Returns a match analysis between the text and the current search.
      Useful for highlighting.
    - `search`: The current search.
    - `setSearch(search)`: Changes the current search. `options` will be automatically filtered.

  code: >
    <Selectable
      onChange={console.log}
      options={[
        { label: 'Daniel Espinoza', value: 2 },
        { label: 'Víctor Sánchez', value: 5 },
        { label: 'Tamara Luque', value: 4 },
        { label: 'Abraham Barrera', value: 1 },
        { label: 'Rodrigo Díaz', value: 3 },
      ]}
      renderer={
        ({
          getOptionProps,
          getSearchMatches,
          options,
          search,
          setSearch,
          value,
        }) =>
          <div className="selectable-dropdown">
            <div className="selectable-dropdown__searchWrapper">
              <input
                className="selectable-dropdown__searchInput"
                onChange={event => setSearch(event.target.value)}
                placeholder="Write to search"
                type="text"
                value={search}
              />
              {search &&
                <button
                  className="selectable-dropdown__searchIcon"
                  onClick={() => setSearch('')}
                  type="button"
                  >
                  &times;
                </button>
              }
            </div>

            {options.map(option =>
              <button
                {...getOptionProps(option, {
                  className: classNames('selectable-dropdown__option', {
                    'selectable-dropdown__option--selected': value === option.value,
                  }),
                  type: 'button',
                })}
                >
                <div>
                  {getSearchMatches(option.label).map(term =>
                    term.match ? <b key={term.fromIndex}>{term.text}</b> : term.text
                  )}
                </div>
              </button>
            )}
          </div>
      }
    />

-
  description: |
    Having a list with a search input is common pattern, so it's already built into Selectable:

    - `getInputProps(otherProps = {})`: Keeps the current search in sync with the input and
      allows to navigate options with the keyboard, using the arrow down, arrow up and enter
      keys.
    - `getClearInputProps(otherProps = {})`: Resets the search and gives the focus back to
      the search input.

    We are also taking advantaje of the virtual focusing that's automatically handled. All we
    need to do is add a class for the option that's currently focused:

    - `focusedElement`: The current focused element `id`. For options this is the index.

  code: >
    <Selectable
      onChange={console.log}
      options={[
        { label: 'Daniel Espinoza', value: 2 },
        { label: 'Víctor Sánchez', value: 5 },
        { label: 'Tamara Luque', value: 4 },
        { label: 'Abraham Barrera', value: 1 },
        { label: 'Rodrigo Díaz', value: 3 },
      ]}
      renderer={
        ({
          focusedElement,
          getClearInputProps,
          getInputProps,
          getOptionProps,
          getSearchMatches,
          options,
          search,
          value,
        }) =>
          <div className="selectable-dropdown">
            <div className="selectable-dropdown__searchWrapper">
              <input
                {...getInputProps({
                  className: 'selectable-dropdown__searchInput',
                  placeholder: 'Write to search',
                  type: 'text',
                })}
              />
              {search &&
                <button
                  {...getClearInputProps({
                    className: 'selectable-dropdown__searchIcon',
                    type: 'button',
                  })}
                  >
                  &times;
                </button>
              }
            </div>

            {options.map((option, index) =>
              <button
                {...getOptionProps(option, {
                  className: classNames('selectable-dropdown__option', {
                    'selectable-dropdown__option--selected': value === option.value,
                    'selectable-dropdown__option--focused': index === focusedElement,
                  }),
                  type: 'button',
                })}
                >
                <div>
                  {getSearchMatches(option.label).map(term =>
                    term.match ? <b key={term.fromIndex}>{term.text}</b> : term.text
                  )}
                </div>
              </button>
            )}
          </div>
      }
    />
