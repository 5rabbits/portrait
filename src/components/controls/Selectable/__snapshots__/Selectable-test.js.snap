// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectable should allow to render options 1`] = `
<ControllableComponent
  options={
    Array [
      Object {
        "label": "Option 1",
        "value": 1,
      },
      Object {
        "label": "Option 10",
        "value": 10,
      },
      Object {
        "label": "Option 5",
        "value": 5,
      },
    ]
  }
  renderer={[Function]}
>
  <Selectable
    blurOnClickOutside={true}
    defaultFocused={false}
    defaultSearch=""
    focused={false}
    focusedElement={undefined}
    onChange={[Function]}
    onFocusedChange={[Function]}
    onFocusedElementChange={[Function]}
    onSearchChange={[Function]}
    options={
      Array [
        Object {
          "label": "Option 1",
          "value": 1,
        },
        Object {
          "label": "Option 10",
          "value": 10,
        },
        Object {
          "label": "Option 5",
          "value": 5,
        },
      ]
    }
    renderer={[Function]}
    search=""
    value={undefined}
  >
    <div
      onMouseLeave={[Function]}
    >
      <div
        key="1"
      >
        Option 1
      </div>
      <div
        key="10"
      >
        Option 10
      </div>
      <div
        key="5"
      >
        Option 5
      </div>
    </div>
  </Selectable>
</ControllableComponent>
`;

exports[`Selectable should render with minimum props 1`] = `
<ControllableComponent
  options={Array []}
  renderer={[Function]}
>
  <Selectable
    blurOnClickOutside={true}
    defaultFocused={false}
    defaultSearch=""
    focused={false}
    focusedElement={undefined}
    onChange={[Function]}
    onFocusedChange={[Function]}
    onFocusedElementChange={[Function]}
    onSearchChange={[Function]}
    options={Array []}
    renderer={[Function]}
    search=""
    value={undefined}
  >
    <span
      onMouseLeave={[Function]}
    />
  </Selectable>
</ControllableComponent>
`;
